File: Poll.py
#!/usr/bin/env python3
##
## EPITECH PROJECT, 2020
## 209Polls
## File description:
## Source file for Poll Class.
##

from sys import argv
from math import sqrt


class Poll:
    def __init__(self, p_size, s_size, p):
        self.var = None
        self.p_size, self.s_size, self.p = p_size, s_size, p
        self.ci95, self.ci99 = 0.0, 0.0

    @staticmethod
    def print_help(exit_code: int = 0):
        print(
            "Usage\n"
            f"\t{argv[0]} pSize sSize p\n\n"
            "DESCRIPTION:\n"
            "\tpSize\tsize of the population\n"
            "\tsSize\tsize of the sample (supposed to be representative)\n"
            "\tp\tpercentage of voting intentions for a specific candidate"
        )
        exit(exit_code)

    def calc_variance(self):
        self.var = ((self.p / 100) * (1 - (self.p / 100)) / self.s_size) * (
                (self.p_size - self.s_size) / (self.p_size - 1))
        self.ci95 = 1.96 * sqrt(self.var) * 100
        self.ci99 = 2.58 * sqrt(self.var) * 100

    def print_results(self):
        print(f"Population size:\t\t{self.p_size}")
        print(f"Sample size:\t\t\t{self.s_size}")
        print(f"Voting intentions:\t\t{self.p}%")
        print(f'Variance:\t\t\t{self.var:.6f}')
        print(f"95% confidence interval:\t[{max(self.p - self.ci95, 0):.2f}%; {min(self.p + self.ci95, 100):.2f}%]")
        print(f"99% confidence interval:\t[{max(self.p - self.ci99, 0):.2f}%; {min(self.p + self.ci99, 100):.2f}%]")

File: bonus/go/poll/Poll.go
package poll

import (
	"fmt"
	"math"
	"sync"
)

const (
	ExitError   = 84
	ExitSuccess = 0
)

type Poll struct {
	Variance     float64
	PSize, SSize int64
	P            float64
	Ci95, Ci99   float64
}

func (p *Poll) CalculateVariance(wg *sync.WaitGroup) {
	p.Variance = ((p.P / 100.0) * (1 - p.P/100.0) / float64(p.SSize)) * (float64(p.PSize-p.SSize) / (float64(p.PSize) - 1.0))
	p.Ci95 = 1.96 * math.Sqrt(p.Variance) * 100
	p.Ci99 = 2.58 * math.Sqrt(p.Variance) * 100
	wg.Done()
}

func (p *Poll) Print() {
	fmt.Printf("Population size:\t\t%d\n", p.PSize)
	fmt.Printf("Sample size:\t\t\t%d\n", p.SSize)
	fmt.Printf("Voting intentions:\t\t%.2f%%\n", p.P)
	fmt.Printf("Variance:\t\t\t%.6f\n", p.Variance)
	fmt.Printf("95%% confidence interval:\t[%.2f%%; %.2f%%]\n",
		math.Max(p.P-p.Ci95, 0), math.Min(p.P+p.Ci95, 100))
	fmt.Printf("99%% confidence interval:\t[%.2f%%; %.2f%%]\n",
		math.Max(p.P-p.Ci99, 0), math.Min(p.P+p.Ci99, 100))
}
